java-tools/OsmAndServerUtilities/src/main/java/net/osmand/server/osmgpx/DownloadOsmGPX.java

```diff
@@ -216,12 +214,197 @@ public static void main(String[] args) throws Exception {
+		if (activitiesMap.isEmpty()) {
+			return null;
+		}
+
+		Map<String, String> tagMap = new LinkedHashMap<>();
+
+		activitiesMap.forEach((activityId, tagList) ->
+				tagList.stream()
+						.sorted((tag1, tag2) -> Integer.compare(tag2.length(), tag1.length()))
+						.forEach(tag -> tagMap.put(tag, activityId))
+		);
+
+		for (Map.Entry<String, String> entry : tagMap.entrySet()) {
+			String tag = entry.getKey();
+			String activityId = entry.getValue();
+			if (tags.contains(tag) || name.toLowerCase().contains(tag) || desc.toLowerCase().contains(tag)) {
+				return activityId;
+			}
+		}
+		return null;
+	}
+
+	private static Map<String, List<String>> createActivitiesMap(String rootPath) {
+		File activityFile = new File(rootPath, "resources/poi/activities.json");
+		if (!activityFile.exists()) {
+			return Collections.emptyMap();
+		}
+
+		ObjectMapper mapper = new ObjectMapper();
+		Map<String, List<String>> activitiesMap = new LinkedHashMap<>();
+
+		try {
+			JsonNode rootNode = mapper.readTree(activityFile);
+
+			JsonNode groups = rootNode.path("groups");
+			for (JsonNode group : groups) {
+				String groupId = group.path("id").asText();
+				List<String> groupTags = new ArrayList<>();
+				JsonNode groupTagsNode = group.path("tags");
+				if (groupTagsNode.isArray()) {
+					for (JsonNode tag : groupTagsNode) {
+						groupTags.add(tag.asText());
+					}
+				}
+				activitiesMap.put(groupId, groupTags);
+				JsonNode activities = group.path("activities");
+				for (JsonNode activity : activities) {
+					String activityId = activity.path("id").asText();
+					List<String> activityTags = new ArrayList<>();
+					JsonNode tagsNode = activity.path("tags");
+					if (tagsNode.isArray()) {
+						for (JsonNode tag : tagsNode) {
+							activityTags.add(tag.asText());
+						}
+					}
+					activitiesMap.put(activityId, activityTags);
+				}
+			}
+		} catch (IOException e) {
+			return Collections.emptyMap();
+		}
+
+		return activitiesMap;
+	}
+
+	private String analyzeActivityFromGpx(byte[] bytes) {
+		GpxFile gpxFile;
+		try (Source src = new Buffer().write(Objects.requireNonNull(Algorithms.gzipToString(bytes)).getBytes())) {
+			gpxFile = GpxUtilities.INSTANCE.loadGpxFile(src);
+		} catch (IOException e) {
+			LOG.error("Error loading GPX file", e);
+			return "error";
+		}
+		if (gpxFile != null) {
+			GpxTrackAnalysis analysis = gpxFile.getAnalysis(System.currentTimeMillis());
+			List<WptPt> points = gpxFile.getAllPoints();
+			if (points.isEmpty() || points.size() < 100 || analysis.getTotalDistance() < 1000 || analysis.getMaxDistanceBetweenPoints() >= 1000) {
+				return "garbage";
+			}
+			float avgSpeed = analysis.getAvgSpeed();
+			if (avgSpeed == 0) {
+				return "other";
+			} else if (avgSpeed <= 12) {
+				return "foot";
+			} else if (avgSpeed <= 25) {
+				return "cycling";
+			} else if (avgSpeed <= 150) {
+				return "driving";
+			} else {
+				return "aviation";
+			}
```

Comment:
Please add validation, for instance: analysis.hasSpeedInTrack() && ... to prevent mark track as "aviation" on case of NaN-like values.